From 78dbca67245e027093a22cc8a9131e142ba3984c Mon Sep 17 00:00:00 2001
From: Aditya Kumar Singh <quic_adisi@quicinc.com>
Date: Fri, 23 Jun 2023 10:13:07 +0530
Subject: [PATCH] ath11k: restrict user country setting during active vdev

Currently, if user wants, country setting can be changed on the fly after
the interface(s) have started beaconing. However, hostapd ignores this regulatory
change event since according to its state, it did not ask for it. This leads
to interfaces which were already active continue beaconing with same old data
which is wrong.

Since there is no potential usecase of changing regulatory on the fly, hence
add change to restrict changing country setting if any of the vdev in it is
active.

Also since regulatory applies to whole SOC, restrict this further to not allow
if any of pdev on a SOC is having an active vdev.

Signed-off-by: Aditya Kumar Singh <quic_adisi@quicinc.com>
---
 drivers/net/wireless/ath/ath11k/reg.c | 38 +++++++++++++++++++++++++++
 1 file changed, 38 insertions(+)

diff --git a/drivers/net/wireless/ath/ath11k/reg.c b/drivers/net/wireless/ath/ath11k/reg.c
index 0a3aaef30957..17a837733cd8 100644
--- a/drivers/net/wireless/ath/ath11k/reg.c
+++ b/drivers/net/wireless/ath/ath11k/reg.c
@@ -43,6 +43,37 @@ static bool ath11k_regdom_changes(struct ath11k *ar, char *alpha2)
 	return memcmp(regd->alpha2, alpha2, 2) != 0;
 }
 
+static bool ath11k_reg_validate_pdev_state(struct ath11k *ar)
+{
+	struct ath11k_base *ab = ar->ab;
+	struct ath11k_pdev *pdev;
+	struct ath11k *tmp_ar;
+	int i;
+
+	rcu_read_lock();
+	for (i = 0; i < ab->num_radios; i++) {
+		pdev = rcu_dereference(ab->pdevs_active[i]);
+		if (!pdev)
+			continue;
+
+		tmp_ar = pdev->ar;
+		if (tmp_ar) {
+			mutex_lock(&tmp_ar->conf_mutex);
+			if (tmp_ar->num_started_vdevs) {
+				if (tmp_ar == ar)
+					ath11k_warn(ab, "%s has active interface, please bring down to set country code",
+						    wiphy_name(ar->hw->wiphy));
+				mutex_unlock(&tmp_ar->conf_mutex);
+				rcu_read_unlock();
+				return false;
+			}
+			mutex_unlock(&tmp_ar->conf_mutex);
+		}
+	}
+	rcu_read_unlock();
+	return true;
+}
+
 static void
 ath11k_reg_notifier(struct wiphy *wiphy, struct regulatory_request *request)
 {
@@ -77,6 +108,13 @@ ath11k_reg_notifier(struct wiphy *wiphy, struct regulatory_request *request)
 		return;
 	}
 
+	/* The SET_INIT_COUNTRY command should not be sent to firmware while any vdev is active.
+	 * Also it does not make sense to give the command for certain pdev's alone.
+	 * Hence check all the pdev's if any have an active vdev before sending the command.
+	 */
+	if (!ath11k_reg_validate_pdev_state(ar))
+		return;
+
 	/* Set the country code to the firmware and will receive
 	 * the WMI_REG_CHAN_LIST_CC EVENT for updating the
 	 * reg info
-- 
2.17.1

